crates/red_knot_module_resolver/src/path.rs

```diff
@@ -325,28 +375,52 @@ impl<'a> ModuleResolutionPathRefInner<'a> {
-                path.extension()
-                    .map_or(true, |ext| ext == "pyi")
-                    .then_some(Self::StandardLibrary(path))
-            }),
-            Self::SitePackages(root) => absolute_path.strip_prefix(root).ok().and_then(|path| {
-                path.extension()
-                    .map_or(true, |ext| matches!(ext, "pyi" | "py"))
-                    .then_some(Self::SitePackages(path))
-            }),
+    fn relativize_path(&self, absolute_path: &FilePathRef<'a>) -> Option<Self> {
+        match (self, absolute_path) {
+            (Self::Extra(root), FilePathRef::System(absolute_path)) => {
+                absolute_path.strip_prefix(root).ok().and_then(|path| {
+                    path.extension()
+                        .map_or(true, |ext| matches!(ext, "py" | "pyi"))
+                        .then_some(Self::Extra(path))
+                })
+            }
+            (Self::FirstParty(root), FilePathRef::System(absolute_path)) => {
+                absolute_path.strip_prefix(root).ok().and_then(|path| {
+                    path.extension()
+                        .map_or(true, |ext| matches!(ext, "pyi" | "py"))
+                        .then_some(Self::FirstParty(path))
+                })
+            }
+            (Self::StandardLibrary(root), FilePathRef::System(absolute_path)) => match root {
+                FilePathRef::System(root) => {
+                    absolute_path.strip_prefix(root).ok().and_then(|path| {
+                        path.extension()
+                            .map_or(true, |ext| ext == "pyi")
+                            .then_some(Self::StandardLibrary(FilePathRef::System(path)))
+                    })
+                }
+                FilePathRef::Vendored(_) => None,
+            },
+            (Self::SitePackages(root), FilePathRef::System(absolute_path)) => {
+                absolute_path.strip_prefix(root).ok().and_then(|path| {
+                    path.extension()
+                        .map_or(true, |ext| matches!(ext, "pyi" | "py"))
+                        .then_some(Self::SitePackages(path))
+                })
+            }
+            (Self::Extra(_), FilePathRef::Vendored(_)) => None,
+            (Self::FirstParty(_), FilePathRef::Vendored(_)) => None,
+            (Self::StandardLibrary(root), FilePathRef::Vendored(absolute_path)) => match root {
+                FilePathRef::System(_) => None,
+                FilePathRef::Vendored(root) => {
+                    absolute_path.strip_prefix(root).ok().and_then(|path| {
+                        path.extension()
+                            .map_or(true, |ext| ext == "pyi")
+                            .then_some(Self::StandardLibrary(FilePathRef::Vendored(path)))
+                    })
+                }
+            },
+            (Self::SitePackages(_), FilePathRef::Vendored(_)) => None,
         }
     }
```

Comment:
I wonder if we could remove a lot of code if the structure would instead be:

```
enum ModuleResolutionPathInner {
    System { path: SystemPath, kind: ResolutionPathKind },
    vendored { path: VendoredPath }
}
```

because most of these branches have exactly the same code. The only difference is the variant taht they construct. But we can defer this for now because I think Carl's recommendation would also address this.